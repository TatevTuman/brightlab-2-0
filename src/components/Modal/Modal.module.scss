.modal {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  z-index: 1;
  transition: all 0.2s ease;

  &[data-animation="true"] {
    animation: 0.2s fade_in;
  }

  @keyframes fade_in {
    from {
      background: green;
    }
    to {
      background: red;
    }
  }

  &-mask {
    position: absolute;
    width: 100%;
    height: 100%;
    background: #000;
    opacity: 0.4;
  }

  &-content {
    width: 60%;
    height: 400px;
    background: #fff;
    z-index: 2;
    border-radius: 10px;

    & > div {
      height: 100%;
    }
  }

  &-header {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
  }

  &-footer {
    margin-top: auto;
    width: 100%;
    height: 20px;
    display: flex;
    align-items: center;
    justify-content: flex-end;
  }
}
