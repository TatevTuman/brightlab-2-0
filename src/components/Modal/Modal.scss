.modal {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  z-index: 1;
  opacity: 0;

  transition: opacity 0.2s ease;
  will-change: opacity;

  &[data-opened='true'] {
    opacity: 1;

    .modal-content {
      transform: translateY(0);
    }
  }

  &[data-opened='false'] {
    opacity: 0;

    .modal-content {
      transform: translateY(-20%);
    }
  }

  &-mask {
    position: absolute;
    width: 100%;
    height: 100%;
    background: #000;
    opacity: 0.4;
  }

  &-content {
    width: 60%;
    height: 400px;
    background: #fff;
    z-index: 2;
    border-radius: 10px;
    transform: translateY(-20%);
    transition: all 0.2s ease;
    will-change: transfrom, translateY;

    & > div {
      height: 100%;
    }
  }

  &-header {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
  }

  &-footer {
    margin-top: auto;
    width: 100%;
    height: 20px;
    display: flex;
    align-items: center;
    justify-content: flex-end;
  }
}
