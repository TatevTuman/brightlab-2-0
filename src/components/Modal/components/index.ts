export { default as ModalHeader } from './ModalHeader/ModalHeader'
export { default as ModalFooter } from './ModalFooter/ModalFooter'

export * from './ModalHeader/ModalHeader'
export * from './ModalFooter/ModalFooter'
