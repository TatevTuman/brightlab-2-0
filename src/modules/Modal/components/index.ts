export { default as ModalHeader } from './ModalHeader/ModalHeader'
export { default as ModalFooter } from './ModalFooter/ModalFooter'

export type { ModalHeaderProps } from './ModalHeader/ModalHeader'
export type { ModalFooterProps } from './ModalFooter/ModalFooter'
