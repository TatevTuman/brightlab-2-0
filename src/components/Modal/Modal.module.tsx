import { LoadableComponent } from '@loadable/component'
import { withLoadableFallback } from '@hocs'
import { ModalFooter, ModalHeader } from './components'

import type { ModalProps, ModalModule } from './Modal'
type LoadableModal = LoadableComponent<ModalProps> & ModalModule

/*
  Modal module
  TODO Research. Dot notation doesn't work with hocs
*/

const Modal = withLoadableFallback<ModalProps>(import('./Modal'), {}) as LoadableModal

Modal.Header = ModalHeader
Modal.Footer = ModalFooter

export default Modal
