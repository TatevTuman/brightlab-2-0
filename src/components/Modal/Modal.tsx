import React from 'react'
import { ModalHeader, ModalFooter, ModalFooterType, ModalHeaderType } from './components'
import { Children } from '@types'
import './Modal.scss'

export interface ModalProps {
  id: string
  opened: boolean
  onClose: () => void
  children: Children
}

export type ModalType = React.FC<ModalProps> & {
  Footer: ModalFooterType
  Header: ModalHeaderType
}

const Modal: ModalType = props => {
  const { id, opened, children, onClose } = props

  if (!opened) return null

  return (
    <div id={id} className={'modal'} role={'complementary'}>
      <div className={'modal-mask'} onClick={onClose} />
      <div className={'modal-content'}>{children}</div>
    </div>
  )
}

Modal.Footer = ModalFooter
Modal.Header = ModalHeader

export default Modal
