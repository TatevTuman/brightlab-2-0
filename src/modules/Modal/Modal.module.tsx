import loadable, { LoadableComponent } from '@loadable/component'
import { withLoadableFallback } from '@hocs'
import { ModalFooter, ModalHeader } from './components'

import type { ModalProps } from './Modal'

export type ModalModule = {
  Footer: typeof ModalHeader
  Header: typeof ModalFooter
}
export type LoadableModal = LoadableComponent<ModalProps> & ModalModule

/*
  Modal module
  TODO Research. Dot notation doesn't work with hocs
*/

const Modal = loadable(() => import('./Modal')) as LoadableModal

Modal.Header = ModalHeader
Modal.Footer = ModalFooter

export default Modal
