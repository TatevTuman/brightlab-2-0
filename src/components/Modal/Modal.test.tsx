import React from 'react'
import { render, cleanup, waitFor, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Button } from '@elements'
import { useModal } from '@hooks'
import Modal from './Modal'

jest.useFakeTimers()

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

const TestModal = () => {
  const { openModal, closeModal } = useModal('TestModal')

  return (
    <>
      <div id={'modals'} data-testid={'modals'} />
      <Modal modalName={'TestModal'}>
        <div data-testid={'modal'}>TestModal</div>
      </Modal>
      <Button onClick={openModal}>Open TestModal</Button>
      <Button onClick={closeModal}>Close TestModal</Button>
    </>
  )
}

describe('Modal', () => {
  it('renders correctly', async () => {
    const { container, getByText, getByTestId } = render(<TestModal />)
    const awaitedContainer = await waitFor(() => container)
    const modalsContainer = getByTestId('modals')

    let modal = awaitedContainer.querySelector('div[data-testid="modal"]')

    expect(modal).toBeNull()
    expect(modalsContainer.childElementCount).toBe(0)

    const openButton = getByText('Open TestModal')
    userEvent.click(openButton)

    modal = awaitedContainer.querySelector('div[data-testid="modal"]')

    expect(modal).toBeInTheDocument()
    expect(modalsContainer.childElementCount).toBe(1)
    expect(awaitedContainer).toMatchSnapshot()

    const closeButton = getByText('Close TestModal')
    userEvent.click(closeButton)

    setTimeout(() => {
      modal = awaitedContainer.querySelector('div[data-testid="modal"]')

      expect(modal).toBeNull()
      expect(modalsContainer.childElementCount).toBe(0)
    }, 1000)

    act(() => {
      jest.runAllTimers()
    })
  })
})
