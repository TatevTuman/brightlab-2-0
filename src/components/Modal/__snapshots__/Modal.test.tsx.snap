// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
<div>
  <div
    data-testid="modals"
    id="modals"
  >
    <div
      class="modal"
      id="TestModal"
      role="complementary"
    >
      <div
        class="modal-mask"
      />
      <div
        class="modal-content"
      >
        <div
          data-testid="modal"
        >
          TestModal
        </div>
      </div>
    </div>
  </div>
  <button
    class="button button__primary-lg "
    data-centered="false"
    data-disabled="false"
    tabindex="0"
    type="button"
  >
    Open TestModal
  </button>
  <button
    class="button button__primary-lg "
    data-centered="false"
    data-disabled="false"
    tabindex="0"
    type="button"
  >
    Close TestModal
  </button>
</div>
`;
